grype/pkg/package.go

```diff
@@ -2,102 +2,257 @@ package pkg
 
 import (
 	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
 
+	"github.com/anchore/grype/internal"
 	"github.com/anchore/grype/internal/log"
 	"github.com/anchore/syft/syft/pkg"
 	"github.com/anchore/syft/syft/source"
 )
 
+// the source-rpm field has something akin to "util-linux-ng-2.17.2-12.28.el6_9.2.src.rpm"
+// in which case the pattern will extract out the following values for the named capture groups:
+//		name = "util-linux-ng"
+//		version = "2.17.2" (or, if there's an epoch, we'd expect a value like "4:2.17.2")
+//		release = "12.28.el6_9.2"
+//		arch = "src"
+var rpmPackageNamePattern = regexp.MustCompile(`^(?P<name>.*)-(?P<version>.*)-(?P<release>.*)\.(?P<arch>[a-zA-Z][^.]+)(\.rpm)$`)
```

Comment:
It seems like there are a lot of type specific things added here; maybe this should predominantly be factored out to the subpackages under [the matcher package](https://github.com/anchore/grype/tree/main/grype/matcher) or something similar?

Reply:
> It seems like there are a lot of type specific things added here

Indeed, this was intentional. The observation was that by moving type specific handling as much as possible upstream in processing (in this case always during package creation) downstream processing becomes simplified.

The reason for the type-specific logic is to craft the `UpstreamPackage` (which is agnostic to the package type) as early as possible. Today (before this PR) we push a lot of that processing into matcher sub-packages, which was causing lots of type assertions to get specific upstream package info out in order to do a fairly standard DB lookup. It became apparent that making the Upstream Package agnostic to the package type (information is inferred from the parent package) made it easier to do this logic when we are creating the Grype Package based off of the Syft package just after SBOM decoding or Syft cataloging (thereby making downstream processing in the matchers simpler). 

I don't think I'd be able to put the rpm-specific logic for parsing into the RPM matcher package since that would introduce a package cycle. However, I could move this logic into another new package, but it isn't clear what that new package's purpose (thus name) would be other than something like `rpm-utils`. For that reason I left this logic unexported in the `pkg` package (closest to where it's used).